Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных:**
1. **Таблица books.** Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.
2. **Таблица authors.** Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.
3. **Таблица publishers.** Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;
4. **Таблица ratings.** Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.
5. **Таблица reviews.** Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

**Задания:**
1. Посчитайте, сколько книг вышло после 1 января 2000 года.
2. Для каждой книги посчитайте количество обзоров и среднюю оценку.
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Шаг 1.** Загрузим библиотеки, получим доступ к базе данных и посмотрим на данные.

**Шаг 2.** Выполним задания, сделав запросы.

**Шаг 3.** Напишем общий вывод.

**Шаг 1. Загрузим библиотеки, получим доступ к базе данных и посмотрим на данные.**

In [41]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Посмотрим на таблицу **books**.

In [42]:
# Выведем первые 5 строк таблицы books и всю информацию
books = '''SELECT * FROM books LIMIT 5'''
desc_query = '''SELECT COUNT (*) FROM books'''
con = engine.connect()
display(pd.io.sql.read_sql(books, con=con))
display(pd.io.sql.read_sql(books, con=con).info())
display(pd.io.sql.read_sql(books, con=con).duplicated().sum())
display(pd.io.sql.read_sql(desc_query, con=con))

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 368.0+ bytes


None

0

count
0   1000

Данные выглядят нормально. Дубликатов нет. Имеет 6 столбцов. Количество строк - 1000.

Посмотрим на таблицу **authors**.

In [43]:
# Выведем первые 5 строк таблицы books и всю информацию
authors = '''SELECT * FROM authors LIMIT 5'''
desc_query1 = '''SELECT COUNT (*) FROM authors'''
con = engine.connect()
display(pd.io.sql.read_sql(authors, con=con))
display(pd.io.sql.read_sql(authors, con=con).info())
display(pd.io.sql.read_sql(authors, con=con).duplicated().sum())
display(pd.io.sql.read_sql(desc_query1, con=con))

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  5 non-null      int64 
 1   author     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


None

0

count
0    636

Данные выглядят нормально. Дубликатов нет. Имеет 2 столбца. Количество строк - 636.

Посмотрим на таблицу **publishers**.

In [44]:
# Выведем первые 5 строк таблицы books и всю информацию
publishers = '''SELECT * FROM publishers LIMIT 5'''
desc_query2 = '''SELECT COUNT (*) FROM publishers'''
con = engine.connect()
display(pd.io.sql.read_sql(publishers, con=con))
display(pd.io.sql.read_sql(publishers, con=con).info())
display(pd.io.sql.read_sql(publishers, con=con).duplicated().sum())
display(pd.io.sql.read_sql(desc_query2, con=con))

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  5 non-null      int64 
 1   publisher     5 non-null      object
dtypes: int64(1), object(1)
memory usage: 208.0+ bytes


None

0

count
0    340

Данные выглядят нормально. Дубликатов нет. Имеет 2 столбца. Количество строк - 340.

Посмотрим на таблицу **ratings**.

In [45]:
# Выведем первые 5 строк таблицы books и всю информацию
ratings = '''SELECT * FROM ratings LIMIT 5'''
desc_query3 = '''SELECT COUNT (*) FROM ratings'''
con = engine.connect()
display(pd.io.sql.read_sql(ratings, con=con))
display(pd.io.sql.read_sql(ratings, con=con).info())
display(pd.io.sql.read_sql(ratings, con=con).duplicated().sum())
display(pd.io.sql.read_sql(desc_query3, con=con))

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   rating     5 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 288.0+ bytes


None

0

count
0   6456

Данные выглядят нормально. Дубликатов нет. Имеет 4 столбца. Количество строк - 6456.

Посмотрим на таблицу **reviews**.

In [46]:
# Выведем первые 5 строк таблицы books и всю информацию
reviews = '''SELECT * FROM reviews LIMIT 5'''
desc_query4 = '''SELECT COUNT (*) FROM reviews'''
con = engine.connect()
display(pd.io.sql.read_sql(reviews, con=con))
display(pd.io.sql.read_sql(reviews, con=con).info())
display(pd.io.sql.read_sql(reviews, con=con).duplicated().sum())
display(pd.io.sql.read_sql(desc_query4, con=con))

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  5 non-null      int64 
 1   book_id    5 non-null      int64 
 2   username   5 non-null      object
 3   text       5 non-null      object
dtypes: int64(2), object(2)
memory usage: 288.0+ bytes


None

0

count
0   2793

Данные выглядят нормально. Дубликатов нет. Имеет 4 столбца. Количество строк - 2793.

**Вывод.**

Мы рассмотрели таблицы, вывели первые строки и информацию о них. Все данные выглядят нормально. Дубликатов нет.

**Шаг 2.** Выполним задания, сделав запросы.

1. Посчитайте, сколько книг вышло после 1 января 2000 года.

In [47]:
count_books = '''
        SELECT COUNT(book_id) 
        FROM books 
        WHERE publication_date > '01.01.2000';
'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(count_books), con = con)

count
0    819

**Вывод**

После 01.01.2000 года вышло **819 книг**.

2. Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [48]:
count_review = '''
       SELECT b.title,
       COUNT (DISTINCT rw.review_id), ROUND(avg(r.rating), 2) as rating
       FROM reviews AS rw
       RIGHT JOIN books AS b ON rw.book_id=b.book_id
       LEFT JOIN ratings AS r ON b.book_id = r.book_id
       GROUP BY b.book_id
       ORDER BY rating DESC
'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(count_review), con = con)

title  count  rating
0        Arrows of the Queen (Heralds of Valdemar  #1)      2    5.00
1    The Walking Dead  Book One (The Walking Dead #...      2    5.00
2                                      Light in August      2    5.00
3    Wherever You Go  There You Are: Mindfulness Me...      2    5.00
4    Captivating: Unveiling the Mystery of a Woman'...      2    5.00
..                                                 ...    ...     ...
995  The World Is Flat: A Brief History of the Twen...      3    2.25
996                  His Excellency: George Washington      2    2.00
997                                      Drowning Ruth      3    2.00
998                                              Junky      2    2.00
999                               Harvesting the Heart      2    1.50

[1000 rows x 3 columns]

**Вывод**.

Посчитали количество обзоров и среднюю оценку за каждую книгу.

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [49]:
best_public = '''
       SELECT p.publisher,
       count(b.book_id) AS total_books
       FROM publishers AS p
       JOIN books AS b ON p.publisher_id = b.publisher_id
       WHERE b.num_pages > 50
       GROUP BY p.publisher
       ORDER BY total_books DESC
'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(best_public), con = con)

publisher  total_books
0                                        Penguin Books           42
1                                              Vintage           31
2                             Grand Central Publishing           25
3                                     Penguin Classics           24
4                                     Ballantine Books           19
..                                                 ...          ...
329                                         Turtleback            1
330  Atheneum Books for Young Readers: Richard Jack...            1
331                                     Penguin Signet            1
332                                    Victor Gollancz            1
333                      Harvard Business Review Press            1

[334 rows x 2 columns]

**Вывод**.

Самое большое количество книг выпустило издательство **Penguin Books**.

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [50]:
best_author = '''
         SELECT
         a.author AS author,
         ROUND(AVG(r.rating), 2) AS avg_rating
         FROM authors AS a
         INNER JOIN books AS b ON b.author_id = a.author_id
         INNER JOIN ratings AS r ON r.book_id = b.book_id
         WHERE b.book_id IN
         (SELECT b.book_id
         FROM books AS b
         JOIN ratings AS r ON b.book_id = r.book_id
         GROUP BY b.book_id
         HAVING COUNT(r.rating) >= 50)
         GROUP BY a.author_id
         ORDER BY avg_rating DESC
'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(best_author), con = con)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré        4.29
1                   Markus Zusak/Cao Xuân Việt Khương        4.26
2                                      J.R.R. Tolkien        4.25
3                                   Louisa May Alcott        4.19
4                                        Rick Riordan        4.08
5                                     William Golding        3.90
6                                       J.D. Salinger        3.83
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.79
8   William Shakespeare/Paul Werstine/Barbara A. M...        3.79
9                                          Lois Lowry        3.75
10                                          Dan Brown        3.74
11            George Orwell/Boris Grabnar/Peter Škerl        3.73
12                                    Stephenie Meyer        3.66
13                                     John Steinbeck        3.62

**Вывод**.

Автор с самой высокой средней оценкой **J.K. Rowling/Mary GrandPré - 4.29**.

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [51]:
avg_reviews = '''
          SELECT
          AVG(count_reviews) AS avg_review
          FROM 
          (SELECT COUNT(review_id) AS count_reviews
          FROM reviews AS r
          WHERE username IN 
          (SELECT username
          FROM ratings 
          GROUP BY username 
          HAVING COUNT(ratings) > 48)
          GROUP BY username) AS review;
'''
con = engine.connect()
pd.io.sql.read_sql(sql = text(avg_reviews), con = con)

avg_review
0        24.0

**Вывод**.

**24 обзора** - среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Шаг 3.** Напишем общий вывод.

**Общий вывод**.

Мы рассмотрели таблицы **books, authors, publishers, ratings, reviews** - вывели первые строки и информацию о них. Все данные выглядят нормально. Дубликатов нет.

Далее провели исследования и ответили на поставленные задачи:
1. После 01.01.2000 года вышло **819 книг**.
2. Посчитали количество обзоров и среднюю оценку за каждую книгу.
3. Определили, что самое большое количество книг выпустило издательство **Penguin Books**.
4. Выявили автора с самой высокой средней оценкой, это **J.K. Rowling/Mary GrandPré - 4.29**.
5. Узнали, что **24 обзора** - среднее количество обзоров от пользователей, которые поставили больше 48 оценок.